File: Session10/127. Word Ladder.java
class Solution {
    public int ladderLength(String beginWord, String endWord, List<String> wordList) {
      Queue<String> q=new LinkedList<>();
      HashSet<String> vis=new HashSet<>(wordList);
      q.add(beginWord);
      int seqLength=0;
       while(!q.isEmpty())
        {
            int s=q.size();
            seqLength+=1;
            for(int i=0;i<s;i++)
            {
                String curr=q.poll();
                for(String nei:findNeighbour(curr,vis))
                {
                    if(isValid(nei,vis))
                    {
                        if(nei.equals(endWord))
                        {
                            return seqLength+1;
                        }
                        vis.remove(nei);
                        q.add(nei);
                    }
                }
            }
        }
        
        return 0;
    }
    
        public boolean isValid(String nei,HashSet<String> vis)
        {
            if(vis.contains(nei))
            {
                return true;
            }
            else
            {
                return false;
            }
        }
 
    
    public List<String> findNeighbour(String curr,HashSet<String> vis)
    {
        char[] curr_arr=curr.toCharArray();
        int n=curr_arr.length;
        List<String> neigh=new ArrayList<String>();
        for(int i=0;i<n;i++)
        {
            char orig= curr_arr[i];
            for(char c='a';c<='z';c++)
            {
               curr_arr[i]=c;
                String nei=String.valueOf(curr_arr );
                if(vis.contains(nei))
                {
                    neigh.add(nei);
                }
            }
            curr_arr[i]=orig;
        }
        return neigh;
    }
}

File: Session10/127. Word Ladder.py
class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        queue = deque([beginWord])
        visited = set([beginWord])
        wordList = set(wordList)
        
        changes = 1
        
        alph = "abcdefghijklmnopqrstuvwxyz"
        
        
        if endWord not in wordList:
            return 0
        
        while queue:
            for l in range(len(queue)):
                curr_word = queue.popleft()
                if curr_word == endWord:
                    return changes

                for i in range(len(curr_word)):
                    prefix, suffix = curr_word[:i], curr_word[i+1:]
                    for al in alph:
                        w = prefix + al + suffix
                        if w in wordList and w not in visited:
                            queue.append(w)
                            visited.add(w)
            changes += 1
            
        return 0